src_c/display.c

```diff
@@ -755,32 +784,45 @@ pg_set_mode(PyObject *self, PyObject *arg, PyObject *kwds)
             else
                 SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 0);
         }
+
 #pragma PG_WARN(Not setting bpp ?)
 #pragma PG_WARN(Add mode stuff.)
+        int x = SDL_WINDOWPOS_UNDEFINED_DISPLAY(display);
+        int y = SDL_WINDOWPOS_UNDEFINED_DISPLAY(display);
+        int w_1=w;
+        int h_1=h;
+
+        if(win){
+            if (SDL_GetWindowDisplayIndex(win) == display) {
+                SDL_GetWindowPosition(win, &x, &y);
+            }
+            if (!(flags & PGS_OPENGL) !=
+                !(SDL_GetWindowFlags(win) & SDL_WINDOW_OPENGL)){
+                SDL_DestroyWindow(win);
+                win=NULL;
+            }
+        }
+
         if (!win) {
             /*open window*/
-            win = SDL_CreateWindow(title,
-                                   SDL_WINDOWPOS_UNDEFINED_DISPLAY(display),
-                                   SDL_WINDOWPOS_UNDEFINED_DISPLAY(display),
-                                   w, h, sdl_flags);
-            if (!win)
-                return RAISE(pgExc_SDLError, SDL_GetError());
-        } else if ((flags & PGS_OPENGL) &&
-                   !(SDL_GetWindowFlags(win) & SDL_WINDOW_OPENGL) ||
-                   !(flags & PGS_OPENGL) &&
-                   (SDL_GetWindowFlags(win) & SDL_WINDOW_OPENGL))
-        {
-            /*recreate existing window*/
-            int x, y, w, h;
-            if (SDL_GetWindowDisplayIndex(win) != display) {
-                x = SDL_WINDOWPOS_UNDEFINED_DISPLAY(display);
-                y = SDL_WINDOWPOS_UNDEFINED_DISPLAY(display);
-            } else {
-                SDL_GetWindowPosition(win, &x, &y);
+            if (flags & PGS_LOGICAL && !(flags & PGS_FULLSCREEN)){
+                SDL_DisplayMode dm;
+                int scale=2;
+
+                if (SDL_GetDesktopDisplayMode(0, &dm) != 0) {
+                    return RAISE(pgExc_SDLError, SDL_GetError());
+                }
+
+                while(w*scale <= dm.w && h*scale <= dm.h){
```

Comment:
There might be a better way to do this scaling: https://stackoverflow.com/a/14731922

Reply:
Sure thing. This could be done in closed form with a couple of divisions, roundings and min() operations, if that makes the code clearer. I don't think it matters much in terms of performance though.